### ISSUE TITLE
Declaration of bit_count shadows a previous local identifier

### ISSUE THREAD

**Issue by tdauth** [facebook/folly#1009]
I get this compilation error (in German):
```
[  1%] Building CXX object CMakeFiles/folly_base.dir/folly/io/async/EventBaseManager.cpp.o
In file included from /home/tdauth/cpp-futures-promises/cmake-build-debug/folly-prefix/src/folly/folly/hash/Checksum.cpp:18:
/home/tdauth/cpp-futures-promises/cmake-build-debug/boost_install/include/boost/crc.hpp: In Elementfunktion »void boost::crc_basic<Bits>::process_bits(unsigned char, std::size_t)«:
/home/tdauth/cpp-futures-promises/cmake-build-debug/boost_install/include/boost/crc.hpp:759:1: Fehler: Deklaration von »bit_count« überdeckt einen vorhergehenden lokalen Bezeichner [-Werror=shadow=compatible-local]
 )
 ^
In file included from /home/tdauth/cpp-futures-promises/cmake-build-debug/boost_install/include/boost/config.hpp:61,
                 from /home/tdauth/cpp-futures-promises/cmake-build-debug/boost_install/include/boost/crc.hpp:12,
                 from /home/tdauth/cpp-futures-promises/cmake-build-debug/folly-prefix/src/folly/folly/hash/Checksum.cpp:18:
/home/tdauth/cpp-futures-promises/cmake-build-debug/boost_install/include/boost/crc.hpp:154:41: Anmerkung: verdeckte Deklaration ist hier
     BOOST_STATIC_CONSTANT( std::size_t, bit_count = Bits );
                                         ^~~~~~~~~
```
As you can see, I am building and installing Boost manually into my build dir ./boost_install and passing the directory when building Folly.
I am using Boost 1.69.0 and Folly v2019.01.28.00 with the flags c++17 and clang++. I had the same error with g++.

**Comment by yfeldblum** [comment]
Is this an issue with folly? Or is this an issue with boost?

**Comment by tdauth** [comment]
I am not sure, I am using the preinstalled Boost version 1.66 now which seems to work but I'd like to compile a custom version of Boost in the future.

**Comment by JoeLoser** [comment]
This is an issue with Boost 1.68 that seems fixed in Boost 1.69. In particular, as seen [on the Boost 1.68 crc code](https://www.boost.org/doc/libs/1_68_0/boost/crc.hpp)

We have a public static member of `crc_basic` with identifier `bit_count`:

```
BOOST_STATIC_CONSTANT( std::size_t, bit_count = Bits );
```
and then a public member function for the same class:

```
template < std::size_t Bits >
void
crc_basic<Bits>::process_bits
(
    unsigned char  bits,
    std::size_t    bit_count // <- this will flag an error if using -Wshadow
)
{
// ..
}
```

The identifier clash of `bit_count` will get flagged when using `-Wshadow`.  But, folly doesn't compile with `-Wshadow` AFAIK based on https://github.com/facebook/folly/blob/master/CMake/FollyCompilerUnix.cmake#L31

I recommend closing this issue. If you wish to apply `-Wshadow` to folly code, you'll want to make it so that you don't also apply that compiler flag to what I would deem system libraries, such as Boost/other third-party code.
